TPCDS connector

petroav opened the pull request.

electrum commented on presto-tpcds/pom.xml:
```diff
@@ -0,0 +1,48 @@
+
+        <dependency>
+            <groupId>com.facebook.presto</groupId>
+            <artifactId>presto-spi</artifactId>
```
This dependency is duplicated

petroav replied:
Done.

kokosing commented:
@petroav Can you please take a why travis fails?

petroav commented:
@kokosing the branch was based on an old SNAPSHOT version so when I pushed a new commit, the SNAPSHOT artifacts were no longer available on OSS Sonatype so the build failed. I just rebased on the latest master and the build passed.

kokosing commented:
@petroav TPC-DS connector uses wrongly VARCHAR where it should use CHAR. Please see for example `STORE` table and column `s_store_id`, when you do `DESCRIBE store;` it will get `VARCHAR(16)` while the tpc-ds specification says it should be `CHAR(16)`. Note tpcds generator uses `character(16)` so the problem is somewhere around connector itself and type mapping.

petroav commented:
@kokosing when we were writing the connector `CHAR(x)` hadn't been merged yet. I just enabled it in https://github.com/prestodb/presto/pull/6374/commits/5ff00214074992641a6ae8533274df239b4437ed.

dain commented on presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsNodePartitioningProvider.java:
```diff
@@ -0,0 +1,82 @@
+import com.facebook.presto.spi.connector.ConnectorTransactionHandle;
+import com.facebook.presto.spi.type.Type;
+import com.google.common.collect.ImmutableMap;
+import sun.reflect.generics.reflectiveObjects.NotImplementedException;
```
`sun.reflect.generics.reflectiveObjects.NotImplementedException`?  Use `UnsupportedOperationException`

petroav replied:
Done.

dain commented on presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsRecordSet.java:
```diff
@@ -0,0 +1,191 @@
+        private List<String> row;
+        private boolean closed;
+
+        public TpcdsRecordCursor(Iterator<List<List<String>>> rows, List<Column> columns)
```
Why is everything a String that is parsed instead of being real raw types like in TPC-H?

petroav replied:
No real reason. I had a look at how TPCH does it and I'll do the appropriate refactors. Changes to the generator are needed so I'll let you know when they're ready.

dain commented on presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsRecordSet.java:
```diff
@@ -0,0 +1,191 @@
+            checkState(row != null, "No current row");
+            Column column = columns.get(field);
+            if (column.getType().getBase() == ColumnType.Base.DATE) {
+                return LocalDate.parse(row.get(column.getPosition())).toEpochDay();
```
`LocalDate` and `LocalTime` are much less efficient than Joda because they create an object with fields and then convert to long where as Joda converts to a long directly.

petroav replied:
I've switched `LocalDate` and `LocalTime` to their Joda equivalents. However, Joda recommends the use of `java.time` as of Java 8. In addition, the Joda equivalent of the above line is a little more awkward:
```
if (column.getType().getBase() == ColumnType.Base.DATE) {
   return Days.daysBetween(new LocalDate(0), LocalDate.parse(row.get(column.getPosition()))).getDays();
}
```
How much do we care about performance here given that this connector won't be used for large scales? I don't feel strongly about this so it's your call.

dain replied:
Last time I checked, the Java 8 versions were a lot slower than Joda when it came to converting dealing with values in ms (long), like we do everywhere in Presto.  In this case I'm more concerned about keeping to the APIs we use in normal Presto code.

dain commented on presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsSplit.java:
```diff
@@ -0,0 +1,129 @@
+    private final boolean noSexism;
+
+    @JsonCreator
+    public TpcdsSplit(@JsonProperty("tableHandle") TpcdsTableHandle tableHandle,
```
place the first argument on a new line

petroav replied:
Done.

dain commented on presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsSplit.java:
```diff
@@ -0,0 +1,129 @@
+    }
+
+    @JsonProperty
+    public boolean isNoSexism()
```
This should be placed with the other getters

petroav replied:
Done.

dain commented on presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsSplit.java:
```diff
@@ -0,0 +1,129 @@
+        TpcdsSplit other = (TpcdsSplit) obj;
+        return Objects.equals(this.tableHandle, other.tableHandle) &&
+                Objects.equals(this.totalParts, other.totalParts) &&
+                Objects.equals(this.partNumber, other.partNumber);
```
Consider adding `noSexism` to equals, hashCode and toString since the setting changes the data

petroav replied:
Done.

dain commented on presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsSplit.java:
```diff
@@ -0,0 +1,129 @@
+        checkState(totalParts >= 1, "totalParts must be >= 1");
+        checkState(totalParts > partNumber, "totalParts must be > partNumber");
+
+        this.tableHandle = requireNonNull(tableHandle, "tableHandle is null");
```
I'd move the null check to be inline with the other checkStates above

petroav replied:
Done.

dain commented on presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsSplit.java:
```diff
@@ -0,0 +1,129 @@
+        this.tableHandle = requireNonNull(tableHandle, "tableHandle is null");
+        this.partNumber = partNumber;
+        this.totalParts = totalParts;
+        this.addresses = ImmutableList.copyOf(addresses);
```
check not null

petroav replied:
Done.

dain commented on presto-tpcds/src/main/java/com/facebook/presto/tpcds/Types.java:
```diff
@@ -0,0 +1,35 @@
+{
+    private Types() {}
+
+    public static <A, B extends A> B checkType(A value, Class<B> target, String name)
```
As I mentioned above, we replaced this in all other connectors with a simple cast.

petroav replied:
Done, class has been deleted.

petroav commented on presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsMetadata.java:
```diff
@@ -0,0 +1,246 @@
+        implements ConnectorMetadata
+{
+    public static final List<String> SCHEMA_NAMES = ImmutableList.of(
+            "sf1", "sf10", "sf100", "sf300", "sf1000", "sf3000", "sf10000", "sf30000", "sf100000");
```
Yep I spoke to @rschlussel and we have some ideas on how to implement this. It'll take some changes to the generator. I'll let you know when this is done.

petroav commented on presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsMetadata.java:
```diff
@@ -0,0 +1,246 @@
+
+        // parse the scale factor
+        int scaleFactor = schemaNameToScaleFactor(tableName.getSchemaName());
+        if (scaleFactor < 0) {
```
Done.

petroav commented on presto-tpcds/src/main/java/com/facebook/presto/tpcds/TpcdsMetadata.java:
```diff
@@ -0,0 +1,246 @@
+            Constraint<ColumnHandle> constraint,
+            Optional<Set<ColumnHandle>> desiredColumns)
+    {
+        TpcdsTableHandle tableHandle = checkType(table, TpcdsTableHandle.class, "table");
```
Done.

petroav commented:
@dain I addressed almost all the comments and rebased to master. Two of the comments require refactors in the generator so I'll re-assign back to you when I'm done.

dain commented:
Looks good.  There are travis failures from:
```
[ERROR] [ERROR] Some problems were encountered while processing the POMs:
[ERROR] 'dependencies.dependency.version' for com.teradata.tpcds:tpcds:jar is missing. @ com.facebook.presto:presto-tpcds:[unknown-version], /home/travis/build/prestodb/presto/presto-tpcds/pom.xml, line 19, column 21
 @ 
```
Once Travis is fixed, can you squash the commits and I'll merge this?

dain commented:
@petroav can you publish the `com.teradata.tpcds:tpcds:jar` so we can land this?

petroav commented:
@dain apologies for the delay on this! @findepi found a bug with the generator at 10TB which I recently fixed (see https://github.com/Teradata/tpcds/issues/19). Didn't make sense to release without fixing that bug. I just finished generating a 30TB dataset and will also generate a 100TB data set to make sure all scales are covered. Right now I'm waiting on Amazon to approve my limit increase request to 100TB.

petroav commented:
@dain I released the generator, rebased and squashed everything. It's ready for you to take a look at.

dain commented:
Anyone can merge this after the current release goes out.

kokosing commented:
Thanks, I will take care of this.

kokosing commented:
@petroav Hey, can you please rebase, also please extract documentation as separate PR as I think @electrum would like to take a look at it before merging. I will merge other commits.

cawallin commented:
@arhimondr has one commit to merge to this branch before it is merged (https://github.com/Teradata/presto/pull/631)

petroav commented:
Closing. This is being superseded by https://github.com/prestodb/presto/pull/8563 and https://github.com/prestodb/presto/pull/8562.
